extension/src/Registry.ts

```diff
@@ -147,17 +145,24 @@ export class Registry {
      */
     public async getPackages(token?: CancellationToken): Promise<Package[]> {
         const packages: Package[] = [];
+        const config = getConfig();
 
         for await (const result of this.findMatchingPackages(this.query, token)) {
             if (token?.isCancellationRequested) {
                 break;
             }
 
             try {
-                let channel = 'latest';
-                if (this.channels && this.channels[result.name]) {
-                    channel = this.channels[result.name];
+                let channel = config.get<any>('channels')?.[result.name] ?? 'latest';
+
+                // Ensure requested channel is valid, otherwise fallback to latest
```

Comment:
Part of the reason for doing the lookup outside of the `getPackageVersionMetadata` was to be able to attach the source channel to the Package instance so that we can (later) show that information in the extension info preview page.

I think that this approach ends up with the same issue that I started with, namely that the resolved package version is trapped inside the `getPackageVersionMetadata` call.

Here is a branch that implements the bulk of the changes you prototyped, but also moves the Package creation inside the `getPackageVersionMetadata`.  We were only using thins call in a few places and in all instances we immediately create the Package instance with the returned manifest.

I pushed another commit for feedback RE the approach.  I'll work on merging with mainline in parallel.